Unit Testing for MyTensor Object Creation in OpenVINO

Saharshjain78 opened the pull request:
### Pull Request Description

This pull request introduces the ability to directly compare `PartialShape` and `Shape` objects with lists and tuples. This enhancement simplifies the process of comparing shapes and dimensions in OpenVINO.

### Changes

#### Shape Class

- Added equality operators in the `Shape` class to support comparison with `std::list<size_t>` and `std::tuple<size_t>`. This allows for direct comparison of a `Shape` object with a list or tuple of dimensions.

- Implemented the `operator[]` and `at()` methods in the `Shape` class to support accessing elements by index. The index is normalized to handle negative indices, similar to Python's indexing.

- Overloaded the `<<` operator for `ov::Shape` to support printing `Shape` objects.

#### PartialShape Class

- Added equality operators in the `PartialShape` class to support comparison with `std::list<Dimension::value_type>` and `std::tuple<Dimension::value_type>`. This allows for direct comparison of a `PartialShape` object with a list or tuple of dimensions.

- Implemented the `operator==` method in `ov::PartialShape` to support comparison of two `PartialShape` objects.

#### Unit Tests

- Added unit tests for the creation, value retrieval, shape checking, and dtype checking of `MyTensor` objects.

These changes aim to improve the usability of the OpenVINO library by simplifying the comparison of shapes and dimensions.

Saharshjain78 commented:
@mlukasze and @akuporos here is the updated pull request to solve this issue https://github.com/openvinotoolkit/openvino/issues/21969 - https://github.com/openvinotoolkit/openvino/pull/25137

akuporos commented:
Hi @Saharshjain78,

Thank you for your contribution!
As it was stated in the original issue, all changes should be performed only for Python API.

> It's need to edit existing/add overloads of `__eq__` for [PartialShape](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/graph/partial_shape.cpp) and [Shape](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/graph/shape.cpp) classes to directly compare with tuple and list.

So please, roll back changes which were made in `src/core/src/partial_shape.cpp` and `src/core/src/shape.cpp` files.
Also please remove excess files from tools folder.

Saharshjain78 commented:
Hi @akuporos ,
I did made the chaanges and made the required changes in the respective files:
Shape: openvino/src/bindings/python/src/pyopenvino/graph/shape.cpp
PartialShape: openvino/src/bindings/python/src/pyopenvino/graph/partial_shape.cpp 
test_core: openvino/src/bindings/python/tests/test_graph/test_core.py
and also reverted the Shape and PartialShape modification.
Please let me know if there are any changes.

Saharshjain78 commented:
Hey @akuporos @mlukasze , I have rolled back the Shape and Partial_Shape, and updated the code of Shape and Partial shape for the python API, the pr needs a review from the maintainer for the further status checks can you please approve the pr or let me know if some changes are to made!

mlukasze commented:
hey @Saharshjain78 thanks for changes, however looks linter is failing - details are attached to the job above

Saharshjain78 commented:
> hey @Saharshjain78 thanks for changes, however looks linter is failing - details are attached to the job above

Hey @mlukasze Can you please guide me a bit more regarding the test failing of linter!
Perhapse you can guide me to a direction!

p-wysocki commented:
Hello @Saharshjain78, you can run linter locally and fix any of the issues using https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/docs/test_examples.md#check-the-codestyle-of-python-api as guide - please let us know if you have any questions regarding that. :) Usually the easiest way to deal with linter errors is to google them if you're not sure what's expected - often other developers provide copy-paste solutions to them.

p-wysocki reviewed (commented):
Thank you for your contribution! The end results look great, the tests are exactly what we need. I left some comments regarding the implementation, please let us know if you have any questions.

Also please remove the two `model_fp32.frozen` files which must have accidentally gotten into the PR.

Saharshjain78 commented:
> Hello @Saharshjain78, you can run linter locally and fix any of the issues using https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/docs/test_examples.md#check-the-codestyle-of-python-api as guide - please let us know if you have any questions regarding that. :) Usually the easiest way to deal with linter errors is to google them if you're not sure what's expected - often other developers provide copy-paste solutions to them.

Sure @p-wysocki thanks for the guidance, I will run it locally to ensure the error is fixed!

akuporos commented:
Hi @Saharshjain78,
There are still two `model_fp32.frozen` files which should be deleted.
We can't merge PR while them are part of it.

akuporos commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -392,6 +392,25 @@ def test_discrete_type_info():
     assert n3.get_type_info().name <= n1.get_type_info().name
 
+def test_shape_equality_with_list():
+    a = [1, 2, 3]
```
Please, use [parametrization](https://docs.pytest.org/en/7.0.x/how-to/parametrize.html)

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,43 @@ void regclass_graph_Shape(py::module m) {
             key += v.size();
         }
-        return v[key];
+        if (key >= 0 && key < static_cast<int64_t>(v.size())) {
+            return v[key];
+        }
+        throw py::index_error("Index out of range");
```
I'm totally ok with `index_arror`.
C++ API errors usually translates to `RuntimeError`, which is too general

Saharshjain78 commented:
Hi @akuporos, I have deleted both the files and the test cases are parametrized and have been updated!

akuporos commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -392,6 +392,23 @@ def test_discrete_type_info():
     assert n3.get_type_info().name <= n1.get_type_info().name
 
+@pytest.mark.parametrize("input_data", [
+    ([1, 2, 3], "list"),
+    ((1, 2, 3), "tuple"),
+])
+def test_shape_equality(input_data):
+    data, _ = input_data
```
```suggestion
    [1, 2, 3],
    (1, 2, 3),
])
def test_shape_equality(input_data):
    data = input_data
```

akuporos commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -392,6 +392,23 @@ def test_discrete_type_info():
+    assert data == s
+
+@pytest.mark.parametrize("input_data", [
+    ([1, 2, 3], "list"),
+    ((1, 2, 3), "tuple"),
+])
+def test_partial_shape_equality(input_data):
+    data, _ = input_data
```
```suggestion
    [1, 2, 3],
    (1, 2, 3),
])
def test_partial_shape_equality(input_data):
    data = input_data
```

akuporos commented:
Hi @Saharshjain78,

Sorry for confusing you,
I see that actually we have these [model_.frozen](https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen) files.
And now you deleted them at all.
Please, return them back (but check that they're not modified like we saw it in your PR).

Saharshjain78 commented:
> Hi @Saharshjain78,
> 
> Sorry for confusing you, I see that actually we have these [model_.frozen](https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen) files. And now you deleted them at all. Please, return them back (but check that they're not modified like we saw it in your PR).

Sure! Here's a warmer and more polished version:

---

Thanks for the clarification! I've added the files in their current state to my PR. I've also updated the test case file with your suggestions for tuples and lists. 
I hope this resolves the issues. If there's anything else, please do let me know!

akuporos commented:
Hi @Saharshjain78,

These model files are changed and it showed in the PR.
Let's try to roll back them.
Please, try these commands for them
`git checkout dd459258029505039c9928a20bed7e6cc0608e6e -- tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen`
`git checkout 5d399faa642cfafa6ee995bf689ad2dfb79bd7e3 -- tools/ovc/unit_tests/moc_tf_fe/test_models/model_fp32.frozen`
And after that just make a commit.

akuporos commented on src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:
```diff
@@ -43,6 +43,21 @@ void regclass_graph_PartialShape(py::module m) {
+    .def("__eq__", [](const ov::PartialShape& self, const py::tuple& tpl) {
+        if (self.rank().is_static() && self.rank().get_length() != tpl.size()) return false;
+        for (size_t i = 0; i < self.rank().get_length(); ++i) {
+            if (!self[i].is_dynamic() && self[i].get_length() != tpl[i].cast<int64_t>()) return false;
```
```suggestion
            if (self[i].is_static() && self[i].get_length() != tpl[i].cast<int64_t>()) return false;
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,28 @@ void regclass_graph_Shape(py::module m) {
+        throw py::index_error("Index out of range");
     });
 
+    py::class_<ov::Shape>(m, "Shape")
+        .def("__eq__", [](const ov::Shape& self, const py::list& lst) {
```
```suggestion
    shape.def("__eq__", [](const ov::Shape& self, const py::list& lst) {
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,28 @@ void regclass_graph_Shape(py::module m) {
+                if (self[i] != lst[i].cast<size_t>()) return false;
+            }
+            return true;
+        })
```
```suggestion
        });
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,28 @@ void regclass_graph_Shape(py::module m) {
+                if (self[i] != lst[i].cast<size_t>()) return false;
+            }
+            return true;
+        })
+        .def("__eq__", [](const ov::Shape& self, const py::tuple& tpl) {
```
```suggestion
        shape.def("__eq__", [](const ov::Shape& self, const py::tuple& tpl) {
```

akuporos replied:
```suggestion
        });
    
        shape.def("__eq__", [](const ov::Shape& self, const py::tuple& tpl) {
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:
```diff
@@ -43,6 +43,21 @@ void regclass_graph_PartialShape(py::module m) {
                        :return: A PartialShape with the given rank (or undefined rank if not provided), and all dimensions are dynamic.
                     )");
 
+    .def("__eq__", [](const ov::PartialShape& self, const py::list& lst) {
```
```suggestion
    shape.def("__eq__", [](const ov::PartialShape& self, const py::list& lst) {
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:
```diff
@@ -43,6 +43,21 @@ void regclass_graph_PartialShape(py::module m) {
+            if (self[i].is_static() && self[i].get_length() != lst[i].cast<int64_t>()) return false;
+        }
+        return true;
+    })
+    .def("__eq__", [](const ov::PartialShape& self, const py::tuple& tpl) {
```
```suggestion
    });
 
    shape.def("__eq__", [](const ov::PartialShape& self, const py::tuple& tpl) {
```

Saharshjain78 commented:
@akuporos I have commited

> Hi @Saharshjain78,
> 
> These model files are changed and it showed in the PR. Let's try to roll back them. Please, try these commands for them `git checkout dd459258029505039c9928a20bed7e6cc0608e6e -- tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen` `git checkout 5d399faa642cfafa6ee995bf689ad2dfb79bd7e3 -- tools/ovc/unit_tests/moc_tf_fe/test_models/model_fp32.frozen` And after that just make a commit.

Hi @akuporos,
I have updated those test files as well as implemented all the changes you suggested in the pull request.

mlukasze commented:
build_jenkins

akuporos commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -392,6 +392,23 @@ def test_discrete_type_info():
+@pytest.mark.parametrize("input_data", [
+    [1, 2, 3],
+    (1, 2, 3),
+])
```
```suggestion
])
@pytest.mark.parametrize("shape_type", [Shape, PartialShape])
```
I'd also prefer to keep one test (as they are the same) and parametrize shape type like this.

akuporos commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -392,6 +392,23 @@ def test_discrete_type_info():
+@pytest.mark.parametrize("input_data", [
+    [1, 2, 3],
+    (1, 2, 3),
+])
+def test_shape_equality(input_data):
+    data = input_data
+    s = Shape(data)
+    assert data == s
```
```suggestion
])
@pytest.mark.parametrize("shape_type", [Shape, PartialShape])
def test_shape_equality(input_data, shape_type):
    data = input_data
    shape = shape_type(data)
    assert data == shape
```
I'd also prefer to keep one test (as they are the same) and parametrize shape type like this.

akuporos commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -392,6 +392,23 @@ def test_discrete_type_info():
+    s = Shape(data)
+    assert data == s
+
+@pytest.mark.parametrize("input_data", [
+    [1, 2, 3],
+    (1, 2, 3),
+])
+def test_partial_shape_equality(input_data):
+    data = input_data
+    ps = PartialShape(data)
+    assert data == ps
```
```suggestion
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,28 @@ void regclass_graph_Shape(py::module m) {
     });
 
+    shape.def("__eq__", [](const ov::Shape& self, const py::list& lst) {
+            if (self.size() != lst.size()) return false;
```
Here the same. I think that common templated function may be created for this logic.

akuporos commented:
Hi @Saharshjain78,

In order to fix C++ code style, please run next commands:
If you don't have installed clang-format-15, so let's install it:
sudo apt --assume-yes install clang-format-15
After that run cmake with -DENABLE_PROFILING_ITT=ON -DSELECTIVE_BUILD=COLLECT in order to enable codestyle check for ITT collector:
cmake -DENABLE_PYTHON=ON -DENABLE_TESTS=ON -DENABLE_PROFILING_ITT=ON -DSELECTIVE_BUILD=COLLECT -B build
And create code style diff:
cmake --build build --target clang_format_fix_all -j8

Saharshjain78 commented:
> Hi @Saharshjain78,
> 
> In order to fix C++ code style, please run next commands: If you don't have installed clang-format-15, so let's install it: sudo apt --assume-yes install clang-format-15 After that run cmake with -DENABLE_PROFILING_ITT=ON -DSELECTIVE_BUILD=COLLECT in order to enable codestyle check for ITT collector: cmake -DENABLE_PYTHON=ON -DENABLE_TESTS=ON -DENABLE_PROFILING_ITT=ON -DSELECTIVE_BUILD=COLLECT -B build And create code style diff: cmake --build build --target clang_format_fix_all -j8

@akuporos I ran the codes (over 1,000 files were changed!) and installed LLVM. Additionally, I updated the `src/bindings/python/src/pyopenvino/graph/partial_shape.cpp` and `src/bindings/python/src/pyopenvino/graph/shape.cpp `files as mentioned, creating a common template.

akuporos commented:
@Saharshjain78 
In OpenVINO we use CLang linter and code style job failed, so please fix it.

praasz commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,30 @@ void regclass_graph_Shape(py::module m) {
             key += v.size();
         }
-        return v[key];
+        if (key >= 0 && key < static_cast<int64_t>(v.size())) {
```
Using v.at(key) will throw if key out of range.
ov::Shape accepts negative index the normalization above is not required.

akuporos commented:
Hi @Saharshjain78,

Is there any chance that you will find a time to finish your PR?

Saharshjain78 commented:
> Hi @Saharshjain78,
> 
> Is there any chance that you will find a time to finish your PR?

Sure I am fixing the linter error

akuporos commented:
Hi @Saharshjain78!

How is your progress?
Do you need any help?

Best regards,
Anastasia

Saharshjain78 commented:
Hi @akuporos ,
I have updated the changes recommended by @praasz!
Just solving the linter error.

praasz commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -44,9 +44,30 @@ void regclass_graph_Shape(py::module m) {
+        throw py::index_error("Index out of range");
+    });
+
+    auto compareShape = [](const ov::Shape& self, const auto& container) {
```
```cpp
    auto compareShape = [](const ov::Shape& self, const auto& container) {
```
This will not work with cpp11 (no generic lambdas),  use template function instead.  Use `snake_case` for functions/variables names.

Consider change this comparator to:
```cpp
return self.size() == container.size() && std::equal(self.begin(), self.end(), container.begin(), bin_predicate);
```

akuporos commented:
Hi @Saharshjain78,

You've added to the commit changes in thirdparties.
Please, roll them back.

Thanx, 
Anastasia

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -41,12 +41,36 @@ void regclass_graph_Shape(py::module m) {
+    auto compareShape = [](const ov::Shape& self, const auto& container) {
+        if (self.size() != container.size()) return false;
+        for (size_t i = 0; i < self.size(); ++i) {
+            if (self[i] != container[i].cast<size_t>()) return false;
```
```suggestion
            if (self[i] != container[i].cast<size_t>()) {
                return false;
            }
```

akuporos replied:
to fix code style job

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -41,12 +41,36 @@ void regclass_graph_Shape(py::module m) {
-        return v[key];
+        // Use at() for bounds checking to throw std::out_of_range if key is out of range
+        try {
+            return v.at(key); // This will automatically check the range
```
```suggestion
            return v.at(key);  // This will automatically check the range
```
to fix code style

akuporos commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -41,12 +41,36 @@ void regclass_graph_Shape(py::module m) {
+    });
+
+    auto compareShape = [](const ov::Shape& self, const auto& container) {
+        if (self.size() != container.size()) return false;
```
```suggestion
        if (self.size() != container.size()) {
            return false;
        }
```

akuporos commented on src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:
```diff
@@ -43,6 +43,22 @@ void regclass_graph_PartialShape(py::module m) {
                     )");
 
+    auto check_shape_equality = [](const ov::PartialShape& self, const auto& container) {
+        if (self.rank().is_static() && self.rank().get_length() != container.size()) return false;
```
```suggestion
        if (self.rank().is_static() && self.rank().get_length() != container.size()) {
            return false;
        }
```

Saharshjain78 reviewed (commented):
Updated the code style as per the required format.

mlukasze commented:
hey @Saharshjain78
clang check is still failing, please take a look and follow with guidlines

p-wysocki commented on src/bindings/python/src/pyopenvino/graph/shape.cpp:
```diff
@@ -41,12 +41,40 @@ void regclass_graph_Shape(py::module m) {
+        return compareShape(self, tpl);
     });
 
+    // Common function to compare shape with list or tuple
+    
+
```
This probably should be deleted. It's also likely why clang formatting check is failing, but the formatting should be applied locally anyway just in case.
```suggestion
```

p-wysocki commented on src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:
```diff
@@ -43,6 +43,26 @@ void regclass_graph_PartialShape(py::module m) {
+        if (self.rank().is_static() && self.rank().get_length() != container.size()) {
+            return false;
+        }
+        for (size_t i = 0; i < self.rank().get_length(); ++i) {
```
I'm not certain this will work for a `PartialShape` with dynamic rank - accessing length of a dynamic dimension will probably cause an error.

Please check and add an appropriate test.

p-wysocki commented on src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:
```diff
@@ -43,6 +43,26 @@ void regclass_graph_PartialShape(py::module m) {
                        :return: A PartialShape with the given rank (or undefined rank if not provided), and all dimensions are dynamic.
                     )");
 
+    auto check_shape_equality = [](const ov::PartialShape& self, const auto& container) {
```
OpenVINO uses C++11, doesn't `auto` as an argument type come with C++14 onwards?

If it does, this needs to be a template util function. Or, since you define this lambda twice in `partial_shape.cpp` and `shape.cpp`, just use explicit type.

akuporos replied:
Please,  use template function instead.

p-wysocki commented on src/bindings/python/tests/test_graph/test_core.py:
```diff
@@ -393,6 +393,17 @@ def test_discrete_type_info():
+    [1, 2, 3],
+    (1, 2, 3),
+])
+@pytest.mark.parametrize("shape_type", [Shape, PartialShape])
```
Please expand `PartialShape` tests with dynamic shapes. That includes:
 - dynamic rank, fully dynamic case
 - static rank, interval dimension
 - static rank, all dimensions dynamic
 - static rank, only some dimensions dynamic

Reference:
https://docs.openvino.ai/2024/api/ie_python_api/_autosummary/openvino.runtime.Dimension.html#openvino.runtime.Dimension.__init__
https://docs.openvino.ai/2024/api/ie_python_api/_autosummary/openvino.runtime.PartialShape.html

p-wysocki reviewed (commented):
Thank you for your contribution! There are some minor changes to be done. Please let us know if you have any questions. :)

Saharshjain78 commented:
Hello @akuporos, @p-wysocki, and @praasz,
I hope you're all doing well. It's been 65 days since I was assigned this issue, and I’ve given it a lot of my time and effort. However, I think it might be best to let someone else take over at this point, as I realize I need more experience to solve it effectively.
Thank you so much for your support throughout this process. I really appreciate it, and I look forward to contributing to other issues in the future. I’m excited about the possibility of working together and having more discussions as I continue to grow and learn.
Thanks again!
